proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,489 @@
+
+Additionally, a constraint extension must not conflict with any other pre-existing extension for the same constraint interface; for the purposes of checking this we use the normal resolution rules. That's what we refer as compiler "coherence".
+
+#### 1. Arguments of the caller function
+
+It looks into the caller function argument list for an evidence of the required extension. Here, `bothValid()` gets a `Validator<A>` passed in so whenever it needs to resolve it for the inner calls to `validate()`, it'll be able to retrieve it from its own argument list.
+```kotlin
+fun <A> validate(a: A, with validator: Validator<A>): Boolean = a.isValid()
+
+fun <A> bothValid(x: A, y: A, with validator: Validator<A>): Boolean = validate(x) && validate(y)
+```
+
+#### 2. Companion object for the target type
+
+In case there's no evidence at the caller function level, we'll look into the companion of the target type. Let's say we have `Repository<A>` and `User` for the `A` type:
+```kotlin
+data class User(val id: Int, val name: String) {
+  companion object {
+    extension class UserValidator(): Validator<User> {
+      override fun User.isValid(): Boolean {
+        return id > 0 && name.length > 0
+      }
+    }
+  }
+}
+```
+That'll be enough for resolving the extension.
+
+#### 3. Companion object for the constraint interface we're looking for
+
+In case there's neither evidence in the companion of the target type, we'll look in the companion of the constraint interface:
+
+```kotlin
+interface Validator<A> {
+  fun A.isValid(): Boolean
+
+  companion object {
+    extension class GroupValidator<A>(with val userValidator: Validator<User>) : Validator<Group> {
+      override fun Group.isValid(): Boolean {
+        for (x in users) {
+          if (!x.isValid()) return false
+        }
+        return true
+      }
+    }
+  }
+}
+```
+
+#### 4. Subpackages of the package where the target type is defined
+
+The next step would be to look into the subpackages of the package where the target type (`User`) is declared.
+
+```kotlin
+package com.domain.repository
+
+import com.domain.User
+
+extension object UserRepository: Repository<User> {
+  override fun loadAll(): List<User> {
+    return listOf(User(25, "Bob"))
+  }
+
+  override fun loadById(id: Int): User? {
+    return if (id == 25) {
+      User(25, "Bob")
+    } else {
+      null
+    }
+  }
+}
+```
+
+Here we got a `Repository<User>` defined in a subpackage of `com.domain`, where the `User` type is defined.
+
+#### 5. Subpackages of the package where the constraint interface is defined
+
+Last place to look at would be subpackages of the package where the constraint interface is defined.
+
+```kotlin
+package com.data.instances
+
+import com.data.Repository
+import com.domain.User
+
+extension object UserRepository: Repository<User> {
+  override fun loadAll(): List<User> {
```

Comment:
Make the implementation use a Map instead.

Reply:
Done in fba1db3 along with another change suggested by @raulraja 👍